File: huffman.py
from __future__ import annotations

from collections import Counter
from dataclasses import dataclass
import heapq
import itertools
from typing import Dict, Optional, Tuple


@dataclass
class Node:
    """Узел дерева Хаффмана.

    Атрибут symbol заполнен только у листьев. Внутренний узел хранит сумму
    частот двух дочерних поддеревьев и используется при декодировании.
    """

    freq: int
    symbol: Optional[int] = None
    left: Optional["Node"] = None
    right: Optional["Node"] = None

    def is_leaf(self) -> bool:
        return self.symbol is not None


def build_tree(data: bytes) -> Optional[Node]:
    """Строит дерево Хаффмана для входной байтовой строки."""
    if not data:
        return None

    # heapq хранит минимальную кучу; второй элемент tuple нужен, чтобы узлы
    # с одинаковыми частотами сравнивались стабильно и без ошибки TypeError.
    counter = itertools.count()
    heap: list[Tuple[int, int, Node]] = []
    for symbol, freq in Counter(data).items():
        heapq.heappush(heap, (freq, next(counter), Node(freq=freq, symbol=symbol)))

    # Если во входе один уникальный байт, дерево состоит из одного листа.
    while len(heap) > 1:
        freq1, _, left = heapq.heappop(heap)
        freq2, _, right = heapq.heappop(heap)
        parent = Node(freq=freq1 + freq2, left=left, right=right)
        heapq.heappush(heap, (parent.freq, next(counter), parent))

    return heap[0][2]


def make_codes(root: Optional[Node]) -> Dict[int, str]:
    """Формирует словарь: байт -> битовая строка."""
    if root is None:
        return {}

    codes: Dict[int, str] = {}

    def walk(node: Node, prefix: str) -> None:
        if node.is_leaf():
            # Для единственного символа нужен хотя бы один бит кода.
            codes[node.symbol] = prefix or "0"
            return
        if node.left is not None:
            walk(node.left, prefix + "0")
        if node.right is not None:
            walk(node.right, prefix + "1")

    walk(root, "")
    return codes


def _pack_bits(bit_string: str) -> tuple[bytes, int]:
    """Упаковывает строку из '0'/'1' в байты и возвращает число полезных битов."""
    bit_length = len(bit_string)
    padding = (-bit_length) % 8
    bit_string += "0" * padding
    out = bytearray()
    for i in range(0, len(bit_string), 8):
        out.append(int(bit_string[i : i + 8], 2))
    return bytes(out), bit_length


def _unpack_bits(payload: bytes, bit_length: int) -> str:
    """Преобразует упакованные байты обратно в битовую строку."""
    return "".join(f"{byte:08b}" for byte in payload)[:bit_length]


def encode(data: bytes) -> dict:
    """Кодирует данные и возвращает учебный контейнер с метаданными."""
    root = build_tree(data)
    codes = make_codes(root)
    bit_string = "".join(codes[byte] for byte in data)
    payload, bit_length = _pack_bits(bit_string)
    return {
        "algorithm": "huffman",
        "codes": {str(k): v for k, v in codes.items()},
        "payload": payload,
        "bit_length": bit_length,
        "original_size": len(data),
    }


def decode(container: dict) -> bytes:
    """Декодирует контейнер, созданный функцией encode."""
    codes = {int(k): v for k, v in container["codes"].items()}
    reverse = {code: symbol for symbol, code in codes.items()}
    bits = _unpack_bits(container["payload"], container["bit_length"])

    current = ""
    out = bytearray()
    for bit in bits:
        current += bit
        if current in reverse:
            out.append(reverse[current])
            current = ""
    return bytes(out)


def compressed_size_bits(container: dict) -> int:
    """Оценивает размер сжатого потока с учётом полезных битов полезной нагрузки."""
    # В промышленном формате нужно компактно хранить дерево/длины кодов.
    # Здесь метаданные хранятся отдельно, поэтому для сравнения алгоритма
    # учитываем битовую длину кодированного сообщения.
    return int(container["bit_length"])

File: compressors/huffman.cpp
#include "huffman.h"
#include <queue>
#include <memory>
#include <bitset>
#include <stdexcept>
#include <algorithm>

struct Node {
    int freq;
    int symbol;
    std::shared_ptr<Node> left;
    std::shared_ptr<Node> right;

    Node(int freq, int symbol = -1, 
         std::shared_ptr<Node> left = nullptr, 
         std::shared_ptr<Node> right = nullptr)
        : freq(freq), symbol(symbol), left(left), right(right) {}

    bool is_leaf() const {
        return symbol != -1;
    }
};

struct CompareNode {
    bool operator()(const std::shared_ptr<Node>& a, const std::shared_ptr<Node>& b) const {
        return a->freq > b->freq;
    }
};

static std::shared_ptr<Node> build_tree(const std::vector<uint8_t>& data) {
    if (data.empty()) return nullptr;
    
    std::vector<int> frequencies(256, 0);
    for (uint8_t byte : data) {
        frequencies[byte]++;
    }
    
    std::priority_queue<std::shared_ptr<Node>, 
                       std::vector<std::shared_ptr<Node>>, 
                       CompareNode> heap;
    
    for (int i = 0; i < 256; i++) {
        if (frequencies[i] > 0) {
            heap.push(std::make_shared<Node>(frequencies[i], i));
        }
    }
    
    if (heap.empty()) return nullptr;
    
    if (heap.size() == 1) {
        return heap.top();
    }
    
    while (heap.size() > 1) {
        auto left = heap.top(); heap.pop();
        auto right = heap.top(); heap.pop();
        auto parent = std::make_shared<Node>(left->freq + right->freq, -1, left, right);
        heap.push(parent);
    }
    
    return heap.top();
}

static void walk_tree(std::shared_ptr<Node> node, const std::string& prefix, 
                     std::vector<std::string>& codes) {
    if (!node) return;
    
    if (node->is_leaf()) {
        codes[node->symbol] = prefix.empty() ? "0" : prefix;
        return;
    }
    
    if (node->left) walk_tree(node->left, prefix + "0", codes);
    if (node->right) walk_tree(node->right, prefix + "1", codes);
}

static std::pair<std::vector<uint8_t>, int> pack_bits(const std::string& bit_string) {
    int bit_length = bit_string.length();
    int padding = (8 - (bit_length % 8)) % 8;
    
    std::string padded_bits = bit_string + std::string(padding, '0');
    std::vector<uint8_t> out;
    out.reserve(padded_bits.length() / 8);
    
    for (size_t i = 0; i < padded_bits.length(); i += 8) {
        uint8_t byte = 0;
        for (size_t j = 0; j < 8 && i + j < padded_bits.length(); j++) {
            if (padded_bits[i + j] == '1') {
                byte |= (1 << (7 - j));
            }
        }
        out.push_back(byte);
    }
    
    return {out, bit_length};
}

static std::string unpack_bits(const std::vector<uint8_t>& payload, int bit_length) {
    std::string result;
    result.reserve(bit_length);
    
    for (size_t i = 0; i < payload.size() && result.length() < (size_t)bit_length; i++) {
        uint8_t byte = payload[i];
        for (int j = 7; j >= 0 && result.length() < (size_t)bit_length; j--) {
            result += ((byte >> j) & 1) ? '1' : '0';
        }
    }
    
    return result;
}

HuffmanContainer Huffman::encode(const std::vector<uint8_t>& data) {
    HuffmanContainer container;
    container.algorithm = "huffman";
    container.original_size = data.size();
    
    if (data.empty()) {
        container.codes = {};
        container.payload = {};
        container.bit_length = 0;
        return container;
    }
    
    auto root = build_tree(data);
    if (!root) {
        container.codes = {};
        container.payload = {};
        container.bit_length = 0;
        return container;
    }
    
    std::vector<std::string> codes(256);
    walk_tree(root, "", codes);
    
    // Сохраняем только ненулевые коды
    for (int i = 0; i < 256; i++) {
        if (!codes[i].empty()) {
            container.codes[i] = codes[i];
        }
    }
    
    std::string bit_string;
    bit_string.reserve(data.size() * 8);
    for (uint8_t byte : data) {
        bit_string += codes[byte];
    }
    
    auto [payload, bit_length] = pack_bits(bit_string);
    container.payload = payload;
    container.bit_length = bit_length;
    
    return container;
}

std::vector<uint8_t> Huffman::decode(const HuffmanContainer& container) {
    if (container.codes.empty() || container.payload.empty()) {
        return {};
    }
    
    // Построение обратного словаря
    std::unordered_map<std::string, int> reverse_codes;
    for (const auto& [symbol, code] : container.codes) {
        reverse_codes[code] = symbol;
    }
    
    std::string bits = unpack_bits(container.payload, container.bit_length);
    
    std::vector<uint8_t> result;
    result.reserve(container.original_size);
    std::string current;
    
    for (char bit : bits) {
        current += bit;
        auto it = reverse_codes.find(current);
        if (it != reverse_codes.end()) {
            result.push_back(static_cast<uint8_t>(it->second));
            current.clear();
        }
    }
    
    return result;
}

int Huffman::compressed_size_bits(const HuffmanContainer& container) {
    return container.bit_length;
}
